x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,324 @@
+import org.elasticsearch.inference.InputType;
+import org.elasticsearch.inference.TaskType;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.inference.action.InferenceAction;
+import org.elasticsearch.xpack.core.inference.results.SparseEmbeddingResults;
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextFieldMapper;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class SemanticQueryBuilder extends AbstractQueryBuilder<SemanticQueryBuilder> {
+    public static final String NAME = "semantic_query";
+
+    private static final ParseField QUERY_FIELD = new ParseField("query");
+
+    private final String fieldName;
+    private final String query;
+
+    private SetOnce<InferenceServiceResults> inferenceResultsSupplier;
+    private InferenceServiceResults inferenceResults;
+
+    public SemanticQueryBuilder(String fieldName, String query) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (query == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + QUERY_FIELD.getPreferredName() + " value");
+        }
+        this.fieldName = fieldName;
+        this.query = query;
+    }
+
+    public SemanticQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.query = in.readString();
+        this.inferenceResults = in.readOptionalNamedWriteable(InferenceServiceResults.class);
+        if (this.inferenceResults != null) {
+            // The supplier is generally not used after the results are set, but set the supplier to maintain equality when serializing
+            // & deserializing
+            this.inferenceResultsSupplier = new SetOnce<>(this.inferenceResults);
+        }
+    }
+
+    private SemanticQueryBuilder(
+        SemanticQueryBuilder other,
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier,
+        InferenceServiceResults inferenceResults
+    ) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+        this.inferenceResults = inferenceResults;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.INFERENCE_FIELDS_METADATA;
```

Comment:
I don't think we need a new transport version for the semantic query, but LMK if I'm missing anything here

Reply:
I think this is needed for the query itself since it's a new one and it needs to serialise over the wire?